File: csharp2/BattleSim.cs
// P113: C# v2 - Async/Await Game State
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

public class Unit
{
    public int Id { get; set; }
    public float Health { get; set; }
    public float Damage { get; set; }
    public float X { get; set; }
    public float Y { get; set; }
}

public class BattleSimulator
{
    private readonly List<Unit> _units = new();
    
    public void AddUnit(Unit unit) => _units.Add(unit);
    
    public float CalculatePower() => _units.Sum(u => u.Health * u.Damage) / 100f;
    
    public Dictionary<int, float> FindThreats()
    {
        var threats = new Dictionary<int, float>();
        foreach (var u in _units)
        {
            int nearby = _units.Count(e => e.Id != u.Id && Distance(u, e) < 50f);
            threats[u.Id] = nearby * 10f;
        }
        return threats;
    }
    
    private static float Distance(Unit a, Unit b)
    {
        float dx = a.X - b.X;
        float dy = a.Y - b.Y;
        return MathF.Sqrt(dx * dx + dy * dy);
    }
    
    public async Task<float> CalculatePowerAsync()
    {
        return await Task.Run(() => CalculatePower());
    }
}

public class Program
{
    public static void Main()
    {
        var sim = new BattleSimulator();
        sim.AddUnit(new Unit { Id = 1, Health = 40, Damage = 5, X = 10, Y = 10 });
        Console.WriteLine($"Power: {sim.CalculatePower()}");
    }
}

File: swift2/BattleSim.swift
// P112: Swift v2 - Actor-based Concurrency
import Foundation

actor BattleSimulator {
    private var units: [Unit] = []
    
    struct Unit: Identifiable {
        let id: Int
        var health: Float
        var damage: Float
        var x: Float
        var y: Float
    }
    
    func addUnit(_ unit: Unit) {
        units.append(unit)
    }
    
    func calculatePower() -> Float {
        let power = units.reduce(Float(0)) { $0 + $1.health * $1.damage }
        return power / 100.0
    }
    
    func findThreats() -> [Int: Float] {
        var threats: [Int: Float] = [:]
        for u in units {
            let nearby = units.filter { unit in
                unit.id != u.id && distance(u, unit) < 50.0
            }.count
            threats[u.id] = Float(nearby) * 10.0
        }
        return threats
    }
    
    private func distance(_ a: Unit, _ b: Unit) -> Float {
        let dx = a.x - b.x
        let dy = a.y - b.y
        return sqrt(dx * dx + dy * dy)
    }
}

Task {
    let sim = BattleSimulator()
    await sim.addUnit(BattleSimulator.Unit(id: 1, health: 40, damage: 5, x: 10, y: 10))
    let power = await sim.calculatePower()
    print("Power: \(power)")
}
